### ISSUE TITLE
Proposal : FilePicker / FolderPicker

### ISSUE THREAD

**Issue by toumir** [xamarin/Essentials#251]
**Description**
**FilePicker** / **FolderPicker** : nice if you add this feature
thanks

> VS bug [#735675](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/735675)

**Comment by AmrAlSayed0** [comment]
This will probably use [ActionGetContent Intent](https://developer.android.com/reference/android/content/Intent#action_get_content) on Android and [UIDocumentPickerViewController](https://developer.apple.com/documentation/uikit/uidocumentpickerviewcontroller) on iOS and [FileOpenPicker](https://docs.microsoft.com/en-us/uwp/api/windows.storage.pickers.fileopenpicker)

Already implemented at [Studyxnet/FilePicker-Plugin-for-Xamarin-and-Windows](https://github.com/Studyxnet/FilePicker-Plugin-for-Xamarin-and-Windows/) but has been abandoned.

**Comment by AmrAlSayed0** [comment]
Since this features seems to be forgotten so I suggest the following API which is exactly the API of [jfversluis/FilePicker-Plugin-for-Xamarin-and-Windows](https://github.com/jfversluis/FilePicker-Plugin-for-Xamarin-and-Windows) a better "fork" of [Studyxnet/FilePicker-Plugin-for-Xamarin-and-Windows](https://github.com/Studyxnet/FilePicker-Plugin-for-Xamarin-and-Windows) :
## FilePicker Class
```csharp
public static classFilePicker
{
    //This returns a long running task which can be awaited for as long as the picker interface is visible. AllowedTypes are optional and defaults to * (all) if not provided.
    Task<FileData> PickFile (string[] allowedTypes = null);
    //Save the file to 
    //Android.OS.Environment.ExternalStorageDirectory on Android,
    //Environment.SpecialFolder.MyDocuments on iOS and
    //ApplicationData.Current.LocalFolder on UWP
    Task<bool> SaveFile (FileData fileToSave);
    //Opens the file using the appropriate document viewer on each platform
    void OpenFile (string fileToOpen);
    //Same as above but takes a FileData class
    void OpenFile (FileData fileToOpen);
}
```
## FileData Class
```csharp
public class FileData : IDisposable
{
    public FileData();
    public FileData(string filePath, string fileName, Func<Stream> streamGetter, Action<bool> dispose = null);

    public byte[] DataArray { get; }
    public string FileName{ get; set; }
    public string FilePath { get; set; }

    public Stream GetStream()
    public void Dispose()
}
```

I think that `SaveFile` and `OpenFile`s are irrelevant here since this is a file *picker* and it should do one thing and that is picking files. For the implementation, it is already done on @jfversluis's repo. I can try and move the code to this repo in my free time but I really have no prior experience with it. Still I will try and report back if things go well.

**Comment by jfversluis** [comment]
That would actually be a great idea. I am trying my best to uphold the Filepicker plugin, but it seems I can't keep up with what is demanded from it. Incorporating it in the Essentials library would be a great addition. I can help out @AmrAlSayed0 but let's wait and see what thoughts are about this move

**Comment by jamesmontemagno** [comment]
Yes, I believe this is feature actually spans multiple potential features that uses files such as camera, email, file picker, etc and all make use of a [File Provider](https://github.com/xamarin/Essentials/issues/130) on Android and a cross-platform File in some way. 

This one is higher up on our list to integrate, but these other points need to go into API design first.

**Comment by vividos** [comment]
@mattleibow @Redth @jamesmontemagno anyone could answer my questions? Thanks!

**Comment by mattleibow** [comment]
Sorry about the delay, for some reason I missed your last comment.

Looking through the issue, I agree with @AmrAlSayed0 about the opening parts being left out - we can actually do that with the Launcher if I am not mistaken. (which is in your obsolete messages) We don't support popping up things from random streams, but we could add that overload.

With regards to letting files out of the app - like share or open other apps - we have a PR that we are working on #416. That will allow us to expose files to the world - just like when you copy it to external storage. Our current work tries to be clever and only copy it if it needs to (and tries to find the best place to copy to)

For docs, @pictos is correct. That wiki should explain things somewhat, but the short of it is write the API first, and once it is done, we run the task and it pops out docs for you to populate.

And finally, and most importantly, the code :) Yup! Just do a PR! All you need to do is create a folder and start working in there.

**Comment by mattleibow** [comment]
@jamesmontemagno @Redth we really need to get #416 looked at and reviewed. It is holding up several features - email, file sharing, sms and anything that needs to expose something inside the app to the world (or other apps).

**Comment by mattleibow** [comment]
Another thing to be aware of since this is working with files/data, is that we have implementations of "file objects" in various PRs - the attachment object in #416 and the media file in #613. Right now, those PR are still open, so you can't really use it, but just go ahead and create a new type 😉 Once we see the requirements for this to work, we can have a look merging all of these together. Chances are, these things will be very similar, but just make sure you have all that you need in your version of the file, and we can discuss. I created a new issue (https://github.com/xamarin/Essentials/issues/685) to track this.

**Comment by jhalbrecht** [comment]
I'm working on learning to TLS secure my mqtt xamForms applications in a real world way for the various platforms.

It occurred to me that an X509Certificate file picker ?and placer? might be handy in xamarin essentials.

**Comment by vividos** [comment]
Until the file picker is added to essentials, you can use jfversluis' FilePicker plugin, it works great!

**Comment by jrahma** [comment]
It should also have the ability to show available service and pick files from them, e.g. Google Drive, Apple Files, DropBox,, etc if installed on device

**Comment by vividos** [comment]
On Android this is already the case; Google Drive and OneDrive offer the files stored online, and surely the other services also. On iOS it's iCloud and OneDrive, maybe others, too.

**Comment by StanleyBroo** [comment]
Any news on this. LOB apps really need this essential control the  FilePicker, to get external files from disk.
I.e. upload a .pdf ..not just images.

**Comment by vividos** [comment]
The Xamarin folks don't react to my messages, sadly. In the meantime you can use this plugin, on which this PR is based on: https://github.com/jfversluis/FilePicker-Plugin-for-Xamarin-and-Windows

**Comment by jrahma** [comment]
This is needed especially when using with [ShareFile](https://docs.microsoft.com/en-us/xamarin/essentials/share?context=xamarin%2Fandroid&tabs=ios#files) +1

**Comment by 1Cor125** [comment]
I see some progress on the dev/file-picker branch. Is this still being worked on? Could this feature use another contributor? Is there a target date for 1.6?
